nixos/modules/security/sgx.nix

```diff
@@ -0,0 +1,167 @@
+        description = "The SGX driver package to use.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    boot.extraModulePackages = [ cfg.packages.driver ];
+    services.udev.packages = [ cfg.packages.psw ];
+
+    # aesmd depends on auditd
+    security.auditd.enable = true;
+
+    # write service config file
+    environment.etc."aesmd.conf".text = cfg.config;
+
+    # create service user and groups
+    users = {
+      users.aesmd = {
+        group = "aesmd";
+        isSystemUser = true;
+        extraGroups = [ "sgx_prv" ];
+        description = "Intel SGX Service Account";
+      };
+
+      groups = {
+        aesmd = { };
+        sgx_prv = { };
+      };
+    };
+
+    systemd.services.aesmd =
+      let
+        path = "${cfg.packages.psw}/aesm";
+      in
+      {
+        description = "Intel(R) Architectural Enclave Service Manager";
+        after = [ "syslog.target" "network.target" "auditd.service" ];
+        wantedBy = [ "multi-user.target" ];
+
+        # restart when config file changes
+        restartTriggers = [ config.environment.etc."aesmd.conf".source ];
+
+        environment = {
+          NAME = "aesm_service";
+          AESM_PATH = "/var/opt/aesmd";
+          LD_LIBRARY_PATH = "${cfg.packages.psw}/lib:${path}";
+        };
+
+        serviceConfig = {
+          ExecStart = "${path}/aesm_service --no-daemon";
+          ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+          ExecStartPre = [ "${pkgs.coreutils}/bin/mkdir -p /var/opt/aesmd/data /var/opt/aesmd/fwdir/data" ];
+
+          Restart = "on-failure";
+          RestartSec = "15s";
+
+          # environment
+          User = "aesmd";
+          Group = "aesmd";
+
+          WorkingDirectory = "/var/opt/aesmd";
+          RuntimeDirectory = "aesmd";
+          RuntimeDirectoryMode = "0755";
+          ReadWritePaths = [ "/var/opt/aesmd" ];
+
+          # Hardening
+          CapabilityBoundingSet = "";
+          DevicePolicy = "closed";
+          DeviceAllow = [
+            "/dev/isgx rw"
+            "/dev/sgx rw"
+            "/dev/sgx/enclave rw"
+            "/dev/sgx/provision rw"
+          ];
+          IPAddressDeny = "any";
+          KeyringMode = "private";
+          LockPersonality = true;
+          MemoryDenyWriteExecute = false;
+          NoNewPrivileges = true;
+          NotifyAccess = "none";
+          ProcSubset = "pid";
+          RemoveIPC = true;
+
+          PrivateDevices = false;
+          PrivateMounts = true;
+          PrivateNetwork = false;
+          PrivateTmp = true;
+          PrivateUsers = true;
+
+          ProtectClock = true;
+          ProtectControlGroups = true;
+          ProtectHome = true;
+          ProtectKernelLogs = true;
+          ProtectKernelModules = true;
+          ProtectKernelTunables = true;
+          ProtectHostname = true;
+          ProtectProc = "invisible";
+          ProtectSystem = "strict";
+          RestrictAddressFamilies = [ "AF_UNIX" ];
```

Comment:
Does aesmd not also accept tcp connections? Does it not also do attestation with Intel over the internet? (Might be wrong about the latter one)

Reply:
I don't believe so, no. See https://github.com/intel/linux-sgx/blob/master/psw/ae/aesm_service/source/core/UnixServerSocket.cpp, it only has a provision for a unix socket.

Reply:
Maybe I am just confused by a docker container I used to run that used tcp? Maybe this was related to the SGX product that was related to this.